### DISCUSSION TITLE
Pagination on multilingual site

### DISCUSSION THREAD

**Discussion by igorfree** [jekyll/jekyll#9578] [Q&A] [answered]
I'm currently building my webpage with jekyll. 

It is supposed to have multiple categories for posts (I want to able to separate stuff like pictures taken on a holiday from a portfolio, etc.), as well as three languages... It's not easy if you have friends all over the world (and hope to reach potential clients all over the world too).

I was able to solve the multilingual aspect quite nicely with collections, as described here: [https://www.kooslooijesteijn.net/blog/multilingual-website-with-jekyll-collections](url). It boils down to creating a collections_dir, with multiple collections in it, one for each language.

But now I want to add pagination. So far, all the guides I found for adding pagination with jekyll-pagination seem to imply that the website has only one language. By that I mean:

- using a single index.html for pagination 
- and single value for pagination_path

correct me if I'm wrong, but it seems to me, this will result in all the new files being written into same subdir. This breaks my existing data structure, which looks like this:

![Screenshot_20240413_193908](https://github.com/jekyll/jekyll/assets/166592551/dcd92d22-cb48-4038-b4f9-3350dd4eaeeb)

Is there a way to have a separate pagination template for each of my collections (languages)?  
I don't want to have posts in English on page 1 and in German on page 2, just because I happened to post couple things in German and then a couple things in English.

**Comment by igorfree** [comment]
I would appreciate all and any help here. Overall, I've been struggling to make jekyll work for me for months, without much success.

What I need:
- multiple languages
- multiple categories of posts (by that I mean, multiple pages that show lists of different posts)
- each category of posts must be paginated

I have read the docs on jekyllrb.com and watched videos on youtube, but nothing works as the information always seems to be either incomplete, ambiguous, or tailored to single-language websites. I can get either one of those things working: multiple languages and categories, OR pagination with just one category and language. 

If I try to use pagination with multiple languages in place, no posts are being displayed. Jekyll does not log any errors and my posts are generated and look fine inside _site folder.

**Reply by fauno** [reply]
It sounds like you need to paginate collections?

https://github.com/sverrirs/jekyll-paginate-v2/blob/master/README-GENERATOR.md#paginating-collections

**Reply by igorfree** [reply]
That's my impression as well. Not only does the pagination not work however, no posts are being displayed when I put 
`
pagination: 
  enabled: true
  collection: [lang]
`
into the front matter

**Comment by igorfree** [comment]
OK, I have figured out, that my data structure was completely wrong and collections were not actually working, because I named the subfolders of "content" without a leading "_".

Now, my multilingual setup is not only working, but actually leveraging collections this time. AFAIK my _posts directory is in the right place, please correct me if I'm wrong. This is how the data structure looks like right now:

![data_structure_correct](https://github.com/jekyll/jekyll/assets/166592551/a46c7a3d-7fd7-4ef2-a3f2-0249cf59343f)

However, after installing, enabling and configuring jekyll-paginate-v2 plugin, when I run `bundle exec jekyll serve --livereload` I get this error message:

`Pagination: Is enabled, but I couldn't find any pagination page. Skipping pagination. Pages must have 'pagination: enabled: true' in their front-matter for pagination to work.`

Yet, all the pages I need pagination on, have the following in their front matter:

`pagination: 
  enabled: true`

So I assumed, the plugin is not finding my pages, because all of them are "hidden" inside the collections. 

Inside _config.yml I then configured the following:

`pagination: ... collection: 'all'`

as to make sure pagination is enabled for all collections. Still same behaivour. 

I see no posts, because the paginator is not finding any pages with enabled pagination (but they exist inside my collections). I just don't have a single page in the root of the website, it's all under the collections and the root of my website only includes an index.html which attempts to guess preferred language and sends the visitor to correct place

**Reply by fauno** [reply]
It looks like _posts should be inside each language subdirectory, like pages are?

Not sure where you're adding the pagination configuration, but they should be on the page that lists the paginated posts/pages, not each individual post/page.

See examples here: https://github.com/sverrirs/jekyll-paginate-v2/tree/master/examples/01-typicalblog

**Reply by igorfree** [reply]
I do have the pagination configuration inside front matter of the page, that has to list some posts. 

blog.md is a page that lists blog posts. In it's front matter I have `pagination: enabled: true`.  

However, paginate-v2 plugin does not recognize blog.md as a page with enabled pagination. See error message above. 

Once I created test.md in the root directory of my website, and added `pagination: enabled: true` to the front matter, paginate-v2 plugin was happy and recognized it as a page with enabled pagination.

However, paginator.posts variable DID NOT include any post. It instead consisted of my PAGES inside content/_[lang]! 

No matter where I put the _posts directory. Inside content, or inside one of the collection, nothing works, paginate-v2 ALWAYS thinks it has to paginate over my pages, instead of my posts.

**Reply by igorfree** [reply]
Allright, it seems the following is the case:

the whole concept of collections interferes with jekylls (or jekyll-paginate-v2s) ability to recognize what is a page and what is a post. 
Once collections are being used, stuff that is inside the collection directory is considered a post and all subdirectories are ignored (at least in my case). If there is a way to let subdirectories also be considered, I have not found it and it won't help anyway, because having pages in the main-navigation AS WELL as inside a list of posts is most definitely not intended.

Here is a solution to this scenario:
- throw away the _posts directory 
- simply put the posts where my pages used to be, inside `content/_[lang]/`
- put the pages in the root directory of the website, atm this means renaming them in [name]_[lang], but plugins like polyglot can help sort that out 
- NOW paginate-v2 ACTUALLY starts doing something useful and allows me to paginate posts from each collection separately.

I suppose once I add in plugins to actually have pre-configured translations in place, it will all become even better and I won't have to create pages multiple times like blog_en.md, blog_de.md etc.

This still leaves the question open where the drafts would go, but this is another story.

Edit: please let me know if there is a better way to do it (aka keep the _posts directory, it seems to be part of basic jekyll design?). Otherwise I'll just mark this as the answer.

**Comment by igorfree** [comment, marked as answer]
so finally, everything works. The actual problem was that guide that I followed. Somehow the link above seems broken, so here it goes again: [problematic multilang guide](https://www.kooslooijesteijn.net/blog/multilingual-website-with-jekyll-collections)

Multilinguality is not covered as such on the jekyll main page, so I just searched online. This is the type of headache that you get for not using official ressources...

When using [polyglot](https://github.com/untra/polyglot) and [paginate-v2](https://github.com/sverrirs/jekyll-paginate-v2) everything works easily and without any problems. Multiple languages, pagination, no misuse of collections for separating your multilingual content and even a key: values list for automatically translating often used words.
